src/Key.php

```diff
@@ -10,13 +10,92 @@ class Key
     /** @var string */
     public $secret;
 
+    /** @var string */
+    public $certificate;
+
+    /** @var string */
+    public $privateKey;
+
+    /** @var string */
+    public $type;
+
     /**
      * @param string $id
      * @param string $secret
      */
-    public function __construct($id, $secret)
+    public function __construct($id, $item)
     {
         $this->id = $id;
-        $this->secret = $secret;
+        $certificate = Key::getX509Certificate($item);
+        $privateKey = Key::getRSAPrivateKey($item);
+        if (($certificate || $privateKey)) {
+            $this->type = 'rsa';
+            if ($privateKey) {
+                $this->privateKey = $privateKey;
+            };
+            if ($certificate) {
+                $this->certificate = openssl_x509_read($certificate);
+            };
+            if ($certificate && $privateKey) {
+                // openssl_pkey_export($this->privateKey, $privateKey);
+            // openssl_x509_export($this->certificate, $certificate);
+            if (! openssl_x509_check_private_key(
+                $this->certificate, $this->privateKey)
+                ) {
+                throw new KeyException("Supplied Certificate and Key are not related");
+            }
+            };
+        } else {
+            $this->type = 'secret';
+            $this->secret = $item;
+            $publicKey = null;
+            $privateKey = null;
+        };
```

Comment:
It looks like you're specifying a very different representation of a key here - I'd suggest the concrete types be split out and implement a similar interface

Reply:
This took me ages to decide on. I wasn't happy just inventing a new architectural component, and burdening the user with having to understand the various key types with explicit calls. This implementation doesn't break existing flows, and automagically sorts the asymmetric types to handle the sign() and verify() flows (which don't differ in the hmac case).

Acknowledged this is not a complete implementation, but is that preferred over compatibility and user experience?

Reply:
If we're looking at an overhaul of the Key type and KeyStore, should we incorporate #18 for the KeyStore itself? Or maybe handle the proposed work under that issue in the future and implement this solution for now?

Reply:
OK so I'd suggest approaching this a little bit differently. Keep the single concrete `Key` and keep it a simple representation of the Key. Ensure your secret is a string only - mixed use variables are a pain IMO. Remove the public variables but expose a
```
public function isRSAPrivateKey()
public function isX509Certificate()
```

Then in `Verification` and `Signature` put the actual logic

Reply:
For RSA there is no one "key", there is a signing key (private) and a verification key (public), so we need to handle these. Keep the internal representation but private, expose them via functions as you describe? Oh, and a certificate is just a container for a public key, we should use/expose it directly for consistency.

```php
public function getRSAPublicKey()
public function getRSAPrivateKey()
public function getHmacSecret()
```

For EC (in the future) add:

```php
public function getECPublicKey()
public function getECPrivateKey()
```

And with key overhaul to handle rotation in #18 (also future):

```php
public function getRSAPublicKeys($keyId)
public function getECPublicKeys($keyId)
public function getHmacSecrets($keyId)
```

Reply:
Generalised to:
```php
public function getSigningKey()
public function getVerifyingKey()
```

Internal state for a key is now private, and the calls return the right public/private or secret for sign/verify. 2ed56e8b30405345f1c08c889490c4909ca61d3f

In future this response can be an array under a single `keyId` to handle key rotation in #18

Reply:
@mtibben, can you review the overhaul? I think it's a lot cleaner and tolerant of different input types (e.g. array of possible verification keys or just a single key).
b9ae276fe63a6bfe3d27fa4d2eb6fa77437746fb